astropy/coordinates/representation/base.py

```diff
@@ -0,0 +1,1600 @@
+            A shallow copy of this representation, without any differentials. If no
+            differentials were present, no copy is made.
+        """
+
+        if not self._differentials:
+            return self
+
+        args = [getattr(self, component) for component in self.components]
+        return self.__class__(*args, copy=False)
+
+    @classmethod
+    def from_representation(cls, representation):
+        """Create a new instance of this representation from another one.
+
+        Parameters
+        ----------
+        representation : `~astropy.coordinates.BaseRepresentation` instance
+            The presentation that should be converted to this class.
+        """
+        return representation.represent_as(cls)
+
+    def __eq__(self, value):
+        """Equality operator for BaseRepresentation
+
+        This implements strict equality and requires that the representation classes are
+        identical, the differentials are identical, and that the representation data are
+        exactly equal.
+        """
+        # BaseRepresentationOrDifferental (checks classes and compares components)
+        out = super().__eq__(value)
+
+        # super() checks that the class is identical so can this even happen? (same
+        # class, different differentials ?)
+        if self._differentials.keys() != value._differentials.keys():
+            raise ValueError("cannot compare: objects must have same differentials")
+
+        for self_diff, value_diff in zip(
+            self._differentials.values(), value._differentials.values()
+        ):
+            out &= self_diff == value_diff
+
+        return out
+
+    def __ne__(self, value):
+        return np.logical_not(self == value)
+
+    def _apply(self, method, *args, **kwargs):
+        """Create a new representation with ``method`` applied to the component
+        data.
+
+        This is not a simple inherit from ``BaseRepresentationOrDifferential`` because
+        we need to call ``._apply()`` on any associated differential classes.
+
+        See docstring for `BaseRepresentationOrDifferential._apply`.
+
+        Parameters
+        ----------
+        method : str or callable
+            If str, it is the name of a method that is applied to the internal
+            ``components``. If callable, the function is applied.
+        *args : tuple
+            Any positional arguments for ``method``.
+        **kwargs : dict
+            Any keyword arguments for ``method``.
+
+        """
+        rep = super()._apply(method, *args, **kwargs)
+
+        rep._differentials = {
+            k: diff._apply(method, *args, **kwargs)
+            for k, diff in self._differentials.items()
+        }
+        return rep
+
+    def __setitem__(self, item, value):
+        if not isinstance(value, BaseRepresentation):
+            raise TypeError(
+                f"value must be a representation instance, " f"not {type(value)}."
```

Comment:
One f-string is good enough for `main`, it should be good enough here as well.
https://github.com/astropy/astropy/blob/87963074a50b14626fbd825536f384a6e96835af/astropy/coordinates/representation.py#L1070